salt: deploy an OCI registry

NicolasT opened the pull request.

bert-e commented:
# Hello nicolast,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @ramanduh

* @alxf

* @tcarmet

* @slaperche-scality

* @cmonfort

* @zempashi

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @alexandre-allard-scality

* @ballot-scality

* @sayf-eddine-scality

* @lucieleonard

* @gdemonet

* @carlito-scality

* @teddyandrieux

NicolasT commented:
TODO:
- [x] Don't run-as-root (see comments in the SLS file)
- [ ] Require credentials for pushing
- [ ] Decrease code complexity (cfr. https://github.com/scality/metalk8s/pull/608#discussion_r257931419)

Over time:
- Proxy behind `nginx` which does TLS termination

NicolasT committed:
vagrant: run `swapoff -a` during Vagrant provisioning

NicolasT committed:
salt: don't check for utilities in `__virtual__`

If we check for utility programs and their functionality in
`__virtual__` of a module, the module will be unavailable when it's
loaded.

However, many of our states install or configure those utilities, and
then want to use a Salt module or state depending on them, which would
then fail.

This commit removes those checks from the various modules. Instead an
invocation of the program would fail at execution time.

NicolasT committed:
salt: add module and state to manage images in a Docker registry

Fixes: #605
Closes: https://github.com/scality/metalk8s/pull/608

NicolasT committed:
salt: inject `pause` image for `containerd`

NicolasT committed:
salt: deploy a Docker registry

Fixes: #602
See: https://github.com/scality/metalk8s/issues/602

NicolasT committed:
salt: populate registry with offline images

NicolasT commented:
This now includes the code from #608, as such closing it.

sayf-eddine-scality commented on salt/metalk8s/registry/populated.sls:
```diff
@@ -0,0 +1,43 @@
+  docker_registry.image_managed:
+    - name: localhost:5000/metalk8s-2.0/{{ image.name }}:{{ image.tag }}
+    - archive_path: /srv/scality/metalk8s-dev/images/{{ image.name }}-{{ image.tag }}.tar.gz
+    - tls_verify: false
```
shouldnt we require pkg: Install skopeo ?

NicolasT replied:
Done.

sayf-eddine-scality commented on salt/metalk8s/registry/populated.sls:
```diff
@@ -0,0 +1,43 @@
+    {
+        'name': 'nginx',
+        'tag': '1.15.8',
+    },
```
time to add salt-master ?

NicolasT replied:
Done.

sayf-eddine-scality commented on salt/metalk8s/registry/init.sls:
```diff
@@ -0,0 +1,49 @@
+  user.present:
+    - name: {{ registry_user }}
+    - fullname: MetalK8s Registry
+    - gid: {{ registry_group }}
```
```
gid_from_name
: False
If True, the default group id will be set to the id of the group with the same name as the user. If the group does not exist the state will fail.
```
Less config data ? :stuck_out_tongue:

NicolasT replied:
It's not configuration, it's a constant defined at the top of this file.

I don't really like `gid_from_name`: I need to use the group name at other places, and I'd rather be explicit about its value instead of using `gid_from_name` here, and then at the other place remember 'Oh, I need the group name here, but should remember that I should use `registry_user` as its value'.

sayf-eddine-scality commented on scripts/bootstrap.sh.in:
```diff
@@ -68,6 +68,8 @@ pillar_roots:
   - module.run
 EOF
+
+        ${SALT_CALL} --local --retcode-passthrough saltutil.sync_all saltenv=metalk8s-@VERSION@
```
we start seeing this kind if calls very often, should we create a wrapper not to rewrite the common arguments every time ?

NicolasT replied:
There's #626

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/oci-registry` and the
destination branch `development/2.0`.

Please resolve the conflict on **the feature branch** (`improvement/oci-registry`).

```bash
 $ git fetch
 $ git checkout origin/improvement/oci-registry
 $ git merge origin/development/2.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/oci-registry
```

NicolasT committed:
salt: import `salt-master` image in Docker registry

See: #602
See: https://github.com/scality/metalk8s/issues/602
See: #606
See: https://github.com/scality/metalk8s/issues/606
See: #610
See: https://github.com/scality/metalk8s/pull/610

NicolasT commented on salt/metalk8s/containerd/configured.sls:
```diff
@@ -1,3 +1,8 @@
+{# One day this should be transferred from the fileserver #}
+{% set pause_image_archive =
+    '/srv/scality/metalk8s-dev/salt/metalk8s/containerd/files/pause-3.1.tar'
```
Hard-coded path for now. Not great, but don't know how to do better for now.

TeddyAndrieux replied:
Why not `file.copy` it somewhere else and then execute.
Because that also mean you can only run this state on the salt master or at least somewhere where you extracted the iso

NicolasT replied:
Ugh, right, I can do this one :blush:

NicolasT replied:
So, yeah, I guess, no, I can't. Reason: at the time of execution, there is no `salt-master`. There is no fileserver. Etc.

NicolasT replied:
So either we need to know whether we're in 'local' mode or in 'normal' mode, and dispatch accordingly, but even then, the source location in 'local' mode needs to be found somehow :(

TeddyAndrieux replied:
My bad that doesn't work with `file.copy` this state is only to copy file localy on the machine but with `file.managed` you copy it even in masterless mode.

NicolasT replied:
I'll give that a try...

NicolasT replied:
@TeddyAndrieux I'd like to get this one in and handle the follow-up through #639

TeddyAndrieux replied:
Ok np

NicolasT commented on salt/metalk8s/registry/populated.sls:
```diff
@@ -0,0 +1,43 @@
+{% set images = [
```
It'd be really good to have this list somewhere else, where the build infrastructure also retrieves it from, etc.

NicolasT commented on salt/metalk8s/registry/populated.sls:
```diff
@@ -0,0 +1,49 @@
+Import {{ image.name }} image:
+  docker_registry.image_managed:
+    - name: localhost:5000/metalk8s-2.0/{{ image.name }}:{{ image.tag }}
+    - archive_path: /srv/scality/metalk8s-dev/images/{{ image.name }}-{{ image.tag }}.tar.gz
```
Same path issue as #639

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alxf

* @jbertran

* @ballot-scality

* @carlito-scality

* @nicolast

* @tcarmet

* @zempashi

* @sayf-eddine-scality

* @gdemonet

* @monpote

* @alexandre-allard-scality

* @lucieleonard

* @slaperche-scality

* @thomasdanan

* @ramanduh

* @nootal

* @teddyandrieux

charlesprost commented on salt/_modules/docker_registry.py:
```diff
@@ -0,0 +1,85 @@
+       This uses the :command:`skopeo` command.
+
+    image
+        Name of the image (e.g. `localhost:5000/metalk8s/nginx:1.14.2`)
```
minor: it's more a URL or URI than a name

NicolasT replied:
I believe `name` is the nomenclature used in Docker spheres. It's definitely not a URL.

charlesprost replied:
`image_path` then?
I'm saying this for the sake of consistency since later on this PR we have:

- `Image destination, e.g. localhost:5000/metalk8s/nginx:1.14.2`
- `Name of the image (including repository), e.g., localhost:5000/metalk8s/nginx:1.14.2.`
- `name: localhost:5000/metalk8s-2.0/{{ image.name }}:{{ image.tag }}`

I fear it could lead to some errors if we do not have the same nomenclature everywhere.

sayf-eddine-scality replied:
i thought we should tag the image with the registry URL to be able to push it no ?

NicolasT replied:
The first part (e.g. `docker.io/` or `localhost:5000/`) is not a URL, the OCI Distribution spec is much more hairy than that. The URL of the repository entrypoint would be `https://docker.io/v2` or `http://localhost:5000/v2`.

charlesprost commented on salt/_modules/docker_registry.py:
```diff
@@ -0,0 +1,85 @@
+    image
+        Name of the image (e.g. `localhost:5000/metalk8s/nginx:1.14.2`)
+    tls_verify : True
+        Verify TLS certificate of the repository. Set to `False` if the
```
It's not clear if the `Set to False` is something automatically performed by the method or if it's an instruction for the developer.

NicolasT replied:
Instruction for the developer. Not sure why it'd be a call argument otherwise?

charlesprost replied:
First time reading it, I was understanding: "This will be automatically set to False if using HTTP because it makes no sense to verify a cert in this situation`. Then I understood the second meaning: "Please set it to False when you use HTTP". I may be the only one making this misunderstanding though.

charlesprost commented on salt/_modules/docker_registry.py:
```diff
@@ -0,0 +1,85 @@
+    destination
+        Image destination, e.g. `localhost:5000/metalk8s/nginx:1.14.2`
+    dest_tls_verify : True
+        Verify TLS certificate of the repository. Set to `False` if the
```
Same comment about the `Set to False`

NicolasT replied:
Same reply as above :P

charlesprost commented on salt/_states/docker_registry.py:
```diff
@@ -0,0 +1,185 @@
+
+
+def __virtual__():
+    if 'docker_registry.get_image_manifest' not in __salt__:
```
Not an expert on Salt, so I may be wrong, but could not we just check for ` if 'docker_registry' not in __salt__:` to see if the module is available? or do we have to check for a specific method?

NicolasT replied:
It's how I saw other states do it.

charlesprost commented on salt/_states/docker_registry.py:
```diff
@@ -0,0 +1,185 @@
+        elif not digest_match and force:
+            if __opts__['test']:
+                return _image_managed_result(
+                    name, None, 'Will force-push archive',
```
s/Will/Would/
```suggestion
                    name, None, 'Would force-push archive',
```

charlesprost commented on salt/_states/docker_registry.py:
```diff
@@ -0,0 +1,185 @@
+                return _image_managed_result(
+                    name, None, 'Will force-push archive',
+                    pchanges=_image_managed_changes(
+                        name, digest, image_manifest['Diges']
```
```suggestion
                        name, digest, image_manifest['Digest']
```

charlesprost commented on salt/_states/docker_registry.py:
```diff
@@ -0,0 +1,185 @@
+    elif image_manifest is None:
+        return _image_managed_new_image(name, archive_path, digest, tls_verify)
+    else:
+        assert False, 'Logic error'
```
There is probably some code to be merged in this file to ease the reading.

NicolasT replied:
I cleaned up quite a bit already after earlier review. I agree, though would like to get the functionality in and rework code if new functionality is required.

charlesprost replied:
Fine by me.

charlesprost reviewed (dismissed):
Apart from a few minor things, seems OK.

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/oci-registry` and the
destination branch `development/2.0`.

Please resolve the conflict on **the feature branch** (`improvement/oci-registry`).

```bash
 $ git fetch
 $ git checkout origin/improvement/oci-registry
 $ git merge origin/development/2.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/oci-registry
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alxf

* @jbertran

* @ballot-scality

* @carlito-scality

* @nicolast

* @tcarmet

* @zempashi

* @sayf-eddine-scality

* @gdemonet

* @monpote

* @alexandre-allard-scality

* @lucieleonard

* @slaperche-scality

* @thomasdanan

* @ramanduh

* @nootal

* @teddyandrieux

NicolasT commented:
@bert-e approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alxf

* @jbertran

* @ballot-scality

* @carlito-scality

* @nicolast

* @tcarmet

* @zempashi

* @sayf-eddine-scality

* @gdemonet

* @monpote

* @alexandre-allard-scality

* @lucieleonard

* @slaperche-scality

* @thomasdanan

* @ramanduh

* @nootal

* @teddyandrieux










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye nicolast.
